<h1 align='center'>PIAIC Class-02</h1>
<br><br><br>
<h3 align='center'>Inroduction to Python</h3>
<br><br><br>
<h4>Date: 24-11-2019</h4>
<h4>Class Time : 12:15 PM </h4>

#### The Zen of Python, by Tim Peters

In [1]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


<h3 align='center'><u>Data Types<u></h3>
    <p>When we assign values to some veriable and want to use that value and its address<br>
        then, we use deep value and shello value techniques....</p>
    <h4> Deep Value and Shello value</h4>

In [5]:
a=8
b=8
print(a)
print(b)

print(id(a))
print(id(b))


8
8
1947709888
1947709888


<p>We can access data directly by using its value and address with veriable name and <i> id()</i> built-in function respectively. <br> Python provide the facility of <b>Memory Management</b> by placing same value that is assigned to different veriables at a single memory address whenever it is not changed.. <br> Above example shows that veriables <i>a</i> and <i>b</i> have same vlaue that is <b>8</b> so this value is going to store at same memory address <b>(1947709888)</b> in storage instead of different memory location that cause excess use of memory which is avoided in <b><i>Python</i></b>.</p>

<h4>Data Type</h4>
<p> To check data type of any veriale python has built-in function <i>type(parameter) </i> </p>


In [12]:
a = 7
print(type(a))
a = '1'
print(type(a))
a = a + '1'    # Adding two values if both string then they concatinate
print(a)
# If we want to convert one veriable data type to another called type casting
b = int(a) + 5
print(type(b))
print(b)

<class 'int'>
<class 'str'>
11
<class 'int'>
16


In [14]:
a = "Monis Ali"
dir(a)         # Want to know which fuction/methods and attributes can be applied to string a

['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getnewargs__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmod__',
 '__rmul__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'capitalize',
 'casefold',
 'center',
 'count',
 'encode',
 'endswith',
 'expandtabs',
 'find',
 'format',
 'format_map',
 'index',
 'isalnum',
 'isalpha',
 'isdecimal',
 'isdigit',
 'isidentifier',
 'islower',
 'isnumeric',
 'isprintable',
 'isspace',
 'istitle',
 'isupper',
 'join',
 'ljust',
 'lower',
 'lstrip',
 'maketrans',
 'partition',
 'replace',
 'rfind',
 'rindex',
 'rjust',
 'rpartition',
 'rsplit',
 'rstrip',
 'split',
 'splitlines',
 'startswith',
 'strip',
 'swapcase',
 'title',
 'translate',
 'upper',
 'zfill']

<b>Operation can be applied by two types</b><br> <b>In-Memory</b><br><b>In-line</b>

In [19]:
a = "MoNiS aLi"
print(a.upper())        # In-line operation
print(a)
a = a.upper()          # In-memory
print(b)

print(a.capitalize())

MONIS ALI
MoNiS aLi
MONIS ALI
Monis ali


In [27]:
a = "       MoNiS aLI        "
print(len(a))
print(a.strip())
b = len(a.strip())
print(b)
print('We saved ',len(a) - b,' Charecters Space/Memory ')

24
MoNiS aLI
9
We saved  15  Charecters Space/Memory 


In [30]:
c = a.split()       # Divide string into list where data is available in strig (Ignore spaces)
print(c)

['MoNiS', 'aLI']


In [34]:
d = " ".join(c)      # Join two or more list' elements
print(d)

MoNiS aLI


In [44]:
a = "We are pakistani we love our country!"
#Slicing : Same as indexing but access more than one vlaues
b = a[7:15:1]                  # veriable[start:end:step] --> end is not included..
print(b)
print(a[::-1])             # Right to left slicing

pakistan
!yrtnuoc ruo evol ew inatsikap era eW


In [53]:
# Search using index
print(a[a.index("pakistani"):16])

pakistani


<b>Result : </b> When we add two strings they get concatinated.... To perform any operation on veriable, operands should have same data type, otherwise <br><b>Type casting </b> should be perform first and then operation.

<b>Lists</b>
It is hytrogenous as list supports to keep different type of data in a list.... <br>


<b> Tuple</b>

<b>Indexing</b>
Left to right indexing starts from 0 to onward and right to left indexing start from -1 to onward...

<b> Sets </b> Same as methimatical sets (Indexing does not support) 

<b> Dectionary</b> It is defined as <br> dectionary = {<br>    <i>Key</i>:<i>Value</i>     <br>} <br><b>Key should be unique</b>

<h4> Concatination </h4><br>

In [55]:
# Multiline string
card = """
    Name    : Monis Ali
    Reg     : 16-CP-45
    Program : PIAIC
"""
print(card)


    Name    : Monis Ali
    Reg     : 16-CP-45
    Program : PIAIC



In [57]:
cards = """
    Name : %s
    Reg : %s
    Program : %s
"""
name = "Monis Ali"
reg = "16-CP-45"
program = "PIAIC"

print(cards%(name,reg,program))


    Name : Monis Ali
    Reg : 16-CP-45
    Program : PIAIC



In [61]:
name = "Monis Ali Qureshi"
reg = "16-CP-45"
program = "PIAIC"

cards = f"""
    Name : {name}
    Reg : {reg}
    Program : {program}
"""

print(cards)


    Name : Monis Ali Qureshi
    Reg : 16-CP-45
    Program : PIAIC



In [70]:
name = "Monis Ali Qureshi"
reg = "16-CP-45"
program = "PIAIC"

cards = """
    Name : {}
    Reg : {}
    Program : {}
"""

print(cards.format(name,reg,program))


    Name : Monis Ali Qureshi
    Reg : 16-CP-45
    Program : PIAIC



In [72]:
name = "Monis Ali Qureshi"
reg = "16-CP-45"
program = "PIAIC"

cards = """
    Name : {a}
    Reg : {b}
    Program : {c}
"""

print(cards.format(a=name,b=reg,c=program))


    Name : Monis Ali Qureshi
    Reg : 16-CP-45
    Program : PIAIC



In [82]:
name = "Monis Ali Qureshi"
reg = "16-CP-45"
program = "PIAIC"

print("Name : " + name + "\nReg : " + reg + "\nProgram : " + program)

Name : Monis Ali Qureshi
Reg : 16-CP-45
Program : PIAIC
